Doxygen/man/man8/arango-upgrade.8
.TH arango-upgrade 8 "So 30. Sep 01:36:14 CEST 2012" "" "ArangoDB"
.SH NAME
arango-upgrade - upgrade script for the ArangoDB database server
.SH SYNOPSIS
arango-upgrade --database.directory database-directory
.SH DESCRIPTION
The arango-upgrade command can be used to perform any necessary system
data modification in an ArangoDB database after a version upgrade.
The command must be run with exclusive access to the server's data directory, 
that means the ArangoDB server must not be running and accessing the
datafiles in the specified data directory.

The arango-upgrade command will print a list of tasks it executes and a 
success/failure status for each. If some task cannot be accomplished
successfully, the root cause must be fixed and the arango-upgrade command be
run again afterwards.
.SH OPTIONS
The arango-upgrade binary the following option:

.IP "--database.directory <string>"
path to the database directory 
.SH EXAMPLES
.EX
shell> arango-upgrade --database.directory /data/arangodb
starts the upgrade script with database directory /data/arangodb 
.EE


.SH AUTHOR
	    Copyright 2012, triAGENS GmbH, Cologne, Germany


CHANGELOG
v1.1.beta1 (2012-XX-XX)
-----------------------

* WARNING:
  - the user has changed from "arango" to "arangodb", the start script has changed from
    "arangod" to "arangodb", the database directory has changed from "/var/arangodb" to
    "/var/lib/arangodb" to be compliant with various Linux policies

  - In 1.1, we have introduced types for collections: regular documents go into document
    collections, and edges go into edge collections. The prefixing (db.xxx vs. edges.xxx) 
    works slightly different in 1.1: edges.xxx can still be used to access collections, 
    however, it will not determine the type of existing collections anymore. To create an 
    edge collection 1.1, you can use db._createEdgeCollection() or edges._create(). 
    And there's of course also db._createDocumentCollection(). 
    db._create() is also still there and will create a document collection by default, 
    whereas edges._create() will create an edge collection.

* the server now handles requests with invalid Content-Length header values as follows:
  - if Content-Length is negative, the server will respond instantly with HTTP 411
    (length required)

  - if Content-Length is positive but shorter than the supplied body, the server will
    respond with HTTP 400 (bad request)

  - if Content-Length is positive but longer than the supplied body, the server will
    wait for the client to send the missing bytes. The server allows 90 seconds for this
    and will close the connection if the client does not send the remaining data

  - if Content-Length is bigger than the maximum allowed size (512 MB), the server will 
    fail with HTTP 413 (request entitiy too large).

  - if the length of the HTTP headers is greated than the maximum allowed size (1 MB),
    the server will fail with HTTP 431 (request header fields too large)

* issue #247: added AQL function MERGE_RECURSIVE

* issue #246: added clear() function in arangosh

* issue #245: Documentation: Central place for naming rules/limits inside ArangoDB

* reduced size of hash index elements by 50 %, allowing more index elements to fit in 
  memory

* issue #235: GUI Shell throws Error:ReferenceError: db is not defined

* issue #229: methods marked as "under construction" 

* issue #228: remove unfinished APIs (/_admin/config/*) 

* having the OpenSSL library installed is now a prerequisite to compiling ArangoDB
  Also removed the --enable-ssl configure option because ssl is always required.

* added AQL functions TO_LIST, NOT_LIST

* issue #224: add optional Content-Id for batch requests

* issue #221: more documentation on AQL explain functionality. Also added
  ArangoStatement.explain() client method

* added db._createStatement() method on server as well (was previously available
  on the client only)

* issue #219: continue in case of "document not found" error in PATHS() function

* issue #213: make waitForSync overridable on specifc actions

* changed AQL optimiser to use indexes in more cases. Previously, indexes might
  not have been used when in a reference expression the inner collection was 
  specified last. Example: FOR u1 IN users FOR u2 IN users FILTER u1._id == u2._id
  Previously, this only checked whether an index could be used for u2._id (not
  possible). It was not checked whether an index on u1._id could be used (possible).
  Now, for expressions that have references/attribute names on both sides of the
  above as above, indexes are checked for both sides.
  
* issue #204: extend the CSV import by TSV and by user configurable 
  seperator character(s)

* issue #180: added support for batch operations

* added startup option --server.backlog-size
  this allows setting the value of the backlog for the listen() system call.
  the default value is 10, the maximum value is platform-dependent

* introduced new configure option "--enable-maintainer-mode" for
  ArangoDB maintainers. this option replaces the previous compile switches
  --with-boost-test, --enable-bison, --enable-flex and --enable-errors-dependency
  the individual configure options have been removed. --enable-maintainer-mode
  turns them all on.

* removed potentially unused configure option --enable-memfail

* fixed issue #197: HTML web interface calls /_admin/user-manager/session

* fixed issue #195: VERSION file in database directory

* fixed issue #193: REST API HEAD request returns a message body on 404

* fixed issue #188: intermittent issues with 1.0.0
  (server-side cursors not cleaned up in all cases, pthreads deadlock issue)

* issue #189: key store should use ISO datetime format bug 

* issue #187: run arango-upgrade on server start 

* fixed issue #183: strange unittest error

* fixed issue #182: manual pages

* fixed issue #181: use getaddrinfo

* moved default database directory to "/var/lib/arangodb" in accordance with 
  http://www.pathname.com/fhs/pub/fhs-2.3.html

* fixed issue #179: strange text in import manual

* fixed issue #178: test for aragoimp is missing

* fixed issue #177: a misleading error message was returned if unknown variables
  were used in certain positions in an AQL query.

* fixed issue #176: explain how to use AQL from the arangosh

* issue #175: re-added hidden (and deprecated) option --server.http-port. This 
  option is only there to be downwards-compatible to Arango 1.0.

* fixed issue #174: missing Documentation for `within`

* fixed issue #170: add db.<coll_name>.all().toArray() to arangosh help screen

* fixed issue #169: missing argument in Simple Queries

* added program arango-upgrade. This program must be run after installing ArangoDB
  and after upgrading from a previous version of ArangoDB. The arango-upgrade script
  will ensure all system collections are created and present in the correct state.
  It will also perform any necessary data updates.

* issue #153: edge collection should be a flag for a collection
  collections now have a type so that the distinction between document and edge 

js/server/arango-upgrade.js
////////////////////////////////////////////////////////////////////////////////
/// @brief arango upgrade actions
///
/// @file
///
/// DISCLAIMER
///
/// Copyright 2004-2012 triAGENS GmbH, Cologne, Germany
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.
///
/// Copyright holder is triAGENS GmbH, Cologne, Germany
///
/// @author Jan Steemann
/// @author Copyright 2012, triAGENS GmbH, Cologne, Germany
////////////////////////////////////////////////////////////////////////////////

// -----------------------------------------------------------------------------
// --SECTION--                                                    arango-upgrade
// -----------------------------------------------------------------------------

////////////////////////////////////////////////////////////////////////////////
/// @addtogroup ArangoDB
/// @{
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
/// @brief updates the database
////////////////////////////////////////////////////////////////////////////////

function main (argv) {
  var internal = require("internal");
  var console = require("console");
  var db = internal.db;

  // path to the VERSION file
  var versionFile = DATABASEPATH + "/VERSION";

  var allTasks = [ ];
  var activeTasks = [ ];
  
  // assume we do not yet have a VERSION value
  var currentVersion = 0;
  if (FS_EXISTS(versionFile)) {
    // VERSION file exists, read its contents
    currentVersion = parseInt(SYS_READ(versionFile));  
  }

  // helper function to define tasks
  function addTask (description, maxVersion, code) {
    // "description" is a textual description of the task that will be printed out on screen
    // "maxVersion" is the maximum version number the task will be applied for
    var task = { description: description, maxVersion: maxVersion, code: code };

    allTasks.push(task);

    // the maximum version number defined for the task is >= than the current VERSION number
    // that means we will apply to task 
    if (currentVersion < parseInt(maxVersion)) {
      activeTasks.push(task);
    }
  }


  // --------------------------------------------------------------------------
  // the actual upgrade tasks. all tasks defined here should be "re-entrant"
  // --------------------------------------------------------------------------

  // set up the collection _users 
  addTask("setup _users collection", 1, function () {
    var users = db._collection("_users");

    if (users == null) {
      users = db._create("_users", { isSystem: true, waitForSync: true });

      if (users == null) {
        console.error("creating users collection '_users' failed");
        return false;
      }
    }

    return true;
  });

  // create a unique index on username attribute in _users
  addTask("create index on username attribute in _users collection", 1, function () {
    var users = db._collection("_users");

    if (users == null) {
      return false;
    }

    users.ensureUniqueConstraint("username");

    return true;
  });
  
  // add a default root user with no passwd
  addTask("add default root user", 1, function () {
    var users = db._collection("_users");

    if (users == null) {
      return false;
    }

    if (users.count() == 0) {
      // only add account if user has not created his/her own accounts already
      users.save({ user: "root", password: internal.encodePassword(""), active: true });
    }

    return true;
  });
  
  // set up the collection _graphs
  addTask("setup _graphs collection", 1, function () {
    var graphs = db._collection("_graphs");

    if (graphs == null) {
      graphs = db._create("_graphs", { isSystem: true, waitForSync: true });

      if (graphs == null) {
        console.error("creating graphs collection '_graphs' failed");
        return false;
      }
    }

    return true;
  });
  
  // create a unique index on name attribute in _graphs
  addTask("create index on name attribute in _graphs collection", 1, function () {
    var graphs = db._collection("_graphs");

    if (graphs == null) {
      return false;
    }

    graphs.ensureUniqueConstraint("name");

    return true;
  });

  // make distinction between document and edge collections
  addTask("set new collection type for edge collections and update collection version", 1, function () {
    var collections = db._collections();
    
    for (var i in collections) {
      var collection = collections[i];

      try {
        if (collection.version() > 1) {
          // already upgraded
          continue;
        }

        if (collection.type() == 3) {
          // already an edge collection
          collection.setAttribute("version", 2);
          continue;
        }
      
        if (collection.count() > 0) {
          var isEdge = true;
          // check the 1st 50 documents from a collection
          var documents = collection.ALL(0, 50);

          for (var j in documents) {
            var doc = documents[j];
  
            // check if documents contain both _from and _to attributes
            if (! doc.hasOwnProperty("_from") || ! doc.hasOwnProperty("_to")) {
              isEdge = false;
              break;
            }
          }

          if (isEdge) {
            collection.setAttribute("type", 3);
            console.log("made collection '" + collection.name() + " an edge collection");
          }
        }
        collection.setAttribute("version", 2);
      }
      catch (e) {
        console.error("could not upgrade collection '" + collection.name() + "'");
        return false;
      }
    }

    return true;
  });
  
  // create the VERSION file
  addTask("create VERSION file", 1, function () {
    // save "1" into VERSION file
    SYS_SAVE(versionFile, "1");
    return true;
  });


  console.log("Upgrade script " + argv[0] + " started");
  console.log("Server VERSION is: " + currentVersion);

  // loop through all tasks and execute them
  console.log("Found " + allTasks.length + " defined task(s), " + activeTasks.length + " task(s) to run");

  var taskNumber = 0;
  for (var i in activeTasks) {
    var task = activeTasks[i];

    console.log("Executing task #" + (++taskNumber) + ": " + task.description);

    var result = task.code();

    if (! result) {
      console.error("Task failed. Aborting upgrade script.");
      console.error("Please fix the problem and try running the upgrade script again.");
      return 1;
    }
    else {
      console.log("Task successful");
    }
  }

  console.log("Upgrade script successfully finished");

  return 0;
}

////////////////////////////////////////////////////////////////////////////////
/// @}
////////////////////////////////////////////////////////////////////////////////

// -----------------------------------------------------------------------------
// --SECTION--                                                       END-OF-FILE
// -----------------------------------------------------------------------------

// Local Variables:
// mode: outline-minor
// outline-regexp: "^\\(/// @brief\\|/// @addtogroup\\|// --SECTION--\\|/// @page\\|/// @}\\)"
// End:
